js/ad_manager_vast.js

```diff
@@ -1875,6 +2127,625 @@ OO.Ads.manager(function(_, $) {
+      currentAd            = null;
+      this.currentAdBeingLoaded = null;
+      this.format               = null;
+      this.node                 = null;
+      adLoaded                  = false;
+    }, this);
+
+    /**
+     * Remove any new lines, line breaks and spaces from string.
+     * @private
+     * @method Vast#_cleanString
+     * @return {string} String with no spaces
+     */
+    var _cleanString = function(string) {
+      return string.replace(/\r?\n|\r/g, '').trim();
+    };
+
+    /**
+     * Check for clickthrough url
+     * @private
+     * @method Vast#_hasClickUrl
+     * @return {object} Ad to look for the clickthrough
+     */
+    var _hasClickUrl = function(ad) {
+      if (ad && ad.data) {
+        var  videoClickTracking = ad.data.videoClickTracking;
+        if (videoClickTracking.clickThrough) {
+          return videoClickTracking.clickThrough.length > 0;
+        }
+      }
+      return false;
+    };
+
+    /**
+     * Check wether or not a vpaid ad is valid by checking the version and the minimum required functions
+     * This is only required for VPAID ads
+     * @method Vast#_isValidVPaid
+     * @private
+     * @return {boolean} VPaid validated value
+     */
+    var _isValidVPaid = _.bind(function() {
+      var creative = currentAd.vpaidAd,
+          vpaidVersion = parseInt(creative.handshakeVersion('2.0')),
+          isValid = true;
+
+      if (vpaidVersion !== 2) {
+        OO.log('VPaid Ad Unit version is not supported.');
+        isValid = false;
+      }
+
+      var requiredFunctions = ['handshakeVersion', 'initAd', 'startAd', 'stopAd', 'skipAd', 'resizeAd',
+                               'pauseAd', 'resumeAd', 'expandAd', 'collapseAd', 'subscribe', 'unsubscribe'];
+      _.each(requiredFunctions, function(fn) {
+        if (!fn && typeof fn !== 'function') {
+          isValid = false;
+          OO.log('VPaid Ad Unit is missing function: ' + fn);
+        }
+      });
+
+      return isValid;
+    }, this);
+
+    /**
+     * Creates a new slot for each ad unit with unique id to avoid conflicts between ads.
+     * This is only required for VPAID ads
+     * @private
+     * @return {object} A DOM element with unique id.
+     */
+    var _createUniqueElement = _.bind(function() {
+      var element = null,
+          parent = this.amc.ui.playerSkinPluginsElement ?
+                              this.amc.ui.playerSkinPluginsElement[0] : this.amc.ui.pluginsElement[0];
+
+      element = document.createElement('div');
+      element.id = _.uniqueId('pluginElement_');
+      element.style.width = '100%';
+      element.style.height = '100%';
+      parent.insertBefore(element, parent.firstChild);
+      return element;
+    }, this);
+
+   /**
+     * Used to generate a frame to load ad media files.
+     * This is only required for VPAID ads
+     * @private
+     */
+    var _getFrame = function() {
+      iframe = document.createElement('iframe');
```

Comment:
Do iframes created by this function get disposed of properly after the ad is finished?